### PR TITLE
Failing test for saving and then loading tokens

### PR DESCRIPTION
**Pull request by christian-marie**
This is something that is happening, not quite sure why yet:

Here is the failing property: https://github.com/anchor/oauth2-server/pull/88/files#diff-a4132e7e065562ab33fba5295a0cce89R147
```
Failures:

  1) Postgres store save then load token
       uncaught exception: ResultError (UnexpectedNull {errSQLType = "varchar", errSQLTableOid = Just (Oid 28910), errSQLField = "client_id", errHaskellType = "ByteString", errMessage = ""}) (after 1 test)
       TokenGrant {grantTokenType = Bearer, grantExpires = 1858-11-17 00:00:00 UTC, grantUsername = Just "nnduqmroqwdjyjzdaeuxvcvaagwxtzrm", grantClientID = Nothing, grantScope = Scope {unScope = fromList ["q"]}}
```

### CODE REVIEW

**Comment by christian-marie** [comment]
Are we meant to be able to attempt to create a TokenGrant without a grantClientId?

**Comment by oswynb** [comment]
Do not believe we can actually, looked through the code we only ever error out or return Just

**Comment by oswynb** [comment]
user id can be empty for certain kinds of grant though (because clients can ask fro grants too I think?)

File: schema/postgresql.sql [hunk 1]
```diff
@@ -27,8 +27,8 @@ CREATE TABLE request_codes (
     authorized   BOOLEAN        NOT NULL DEFAULT FALSE,
     expires TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW() + '10 minutes',
 
-    user_id      VARCHAR(256)   NOT NULL,
-    client_id    VARCHAR(37)    NOT NULL,
+    user_id      VARCHAR(256),
+    client_id    VARCHAR(37),
```
Comment (oswynb) [inline schema/postgresql.sql]: So user_id is probably optional, but pretty sure client_id must be not null

**Comment by oswynb** [comment]
Apart from that one point, rest looks good

**Comment by christian-marie** [comment]
Cool, will fix. thanks.

**Comment by tvh** [comment]
I would say `client_id` cannot be null. This means, that you have to register a client even if you are just hacking on a project. This is what most big oauth2 deployments go. Github is an exception here, but I like the idea that we know all the clients and can revoke them if necessary.

`user_id` is a bit trickier. If we want to have `service` applications common to the whole infrastructure, then this has to be nullable. For every other request however it should contain the user_id.

File: test/tokenstore.hs [hunk 2]
```diff
@@ -60,18 +60,24 @@ instance Arbitrary Scope where
...
-    (^?! packedChars . token) <$> vectorOf 63 alphabet
+    arbitrary =
+        (^?! packedChars . token) <$> vectorOf 63 alphabet
 
 instance Arbitrary TokenType where
     arbitrary = arbitraryBoundedEnum
 
 instance Arbitrary TokenGrant where
-    arbitrary =
+    arbitrary = do
```
Comment (oswynb) [inline test/tokenstore.hs]: Why the extra do?

**Pull request merged**